src/zero_copy/text.rs

```diff
@@ -59,32 +95,297 @@ impl Char for u8 {
+where
+    C: Char,
+    I: StrInput<C> + ?Sized,
+    E: Error<I>,
+{
+    any()
+        .filter(move |c: &C| c.is_digit(radix))
+        .repeated()
+        .at_least(1)
+        .map_slice(|x| x)
+}
+
+/// A parser that accepts a non-negative integer.
+///
+/// An integer is defined as a non-empty sequence of ASCII digits, where the first digit is non-zero or the sequence
+/// has length one.
+///
+/// The output type of this parser is [`Character::Collection`] (i.e: [`String`] when `C` is [`char`], and [`Vec<u8>`]
+/// when `C` is [`u8`]).
+///
+/// The `radix` parameter functions identically to [`char::is_digit`]. If in doubt, choose `10`.
+///
+/// # Examples
+///
+/// ```
+/// # use chumsky::zero_copy::prelude::*;
+/// let dec = text::int::<_, _, Simple<str>>(10)
+///     .then_ignore(end());
+///
+/// assert_eq!(dec.parse("0").0, Some("0"));
+/// assert_eq!(dec.parse("1").0, Some("1"));
+/// assert_eq!(dec.parse("1452").0, Some("1452"));
+/// // No leading zeroes are permitted!
+/// assert!(dec.parse("04").0.is_none());
+///
+/// let hex = text::int::<_, _, Simple<str>>(16)
+///     .then_ignore(end());
+///
+/// assert_eq!(hex.parse("2A").0, Some("2A"));
+/// assert_eq!(hex.parse("d").0, Some("d"));
+/// assert_eq!(hex.parse("b4").0, Some("b4"));
+/// assert!(hex.parse("0B").0.is_none());
+/// ```
+///
+#[must_use]
+pub fn int<'a, I: StrInput<C> + ?Sized, C: Char, E: Error<I>>(
+    radix: u32,
+) -> impl Parser<'a, I, &'a C::Slice, E> {
+    any()
+        .filter(move |c: &C| c.is_digit(radix) && c != &C::digit_zero())
+        .map(Some)
+        .then(any().filter(move |c: &C| c.is_digit(radix)).repeated())
+        .ignored()
+        .or(just(C::digit_zero()).ignored())
+        .map_slice(|x| x)
+}
+
+/// A parser that accepts a C-style identifier.
+///
+/// The output type of this parser is [`Character::Collection`] (i.e: [`String`] when `C` is [`char`], and [`Vec<u8>`]
+/// when `C` is [`u8`]).
+///
+/// An identifier is defined as an ASCII alphabetic character or an underscore followed by any number of alphanumeric
+/// characters or underscores. The regex pattern for it is `[a-zA-Z_][a-zA-Z0-9_]*`.
+#[must_use]
+pub fn ident<'a, I: StrInput<C> + ?Sized, C: Char, E: Error<I>>(
+) -> impl Parser<'a, I, &'a C::Slice, E> {
+    any()
+        .filter(|c: &C| c.to_char().is_ascii_alphabetic() || c.to_char() == '_')
+        .then(
+            any()
+                .filter(|c: &C| c.to_char().is_ascii_alphanumeric() || c.to_char() == '_')
+                .repeated(),
+        )
+        .map_slice(|x| x)
+}
+
+/// A parser that consumes text and generates tokens using semantic whitespace rules and the given token parser.
+///
+/// Also required is a function that collects a [`Vec`] of tokens into a whitespace-indicated token tree.
+#[must_use]
+pub fn semantic_indentation<'a, Tok, T, F, E: Error<str> + 'a>(
+    token: T,
+    make_group: F,
+) -> impl Parser<'a, str, Vec<Tok>, E> + 'a
```

Comment:
I think this function is fine to keep for now, but I'd really like to add much better support for 'parental context' parsers. IIRC this function hasn't been published to crates.io yet, so I might just remove it before the next version in favour of something more powerful.

Reply:
I've added it because my parser for my type checker uses it and I want to switch to zero-copy and also to make the pythonic example work with zero_copy for whatever reason?

Reply:
That's fair enough. We can keep it until we have something that's at least as powerful as it as an alternative :)